Add assertions for language objects and text direction (issue 598)

mmccool opened the pull request:
- Be specific about where `@language` tag can appear to simplify processing (this is an extension of issue noted below...)
- Add assertions to indicate how base text direction is to be inferred from default language outside of MultiLanguage contexts, including requirement for script subtags if necessary (btw, an example of the latter is az-Latn vs az-Arab)
- Make LTR the default text direction if no default language is given
- Script subtag required in MultiLanguage.  I also moved the assertions regarding language tags from ontology/td.ttl to validation/td-validation.ttl to be consistent with other such assertions, and marked them up with rfc2119-assertion spans.

Addresses Issue https://github.com/w3c/wot-thing-description/issues/598

mkovatsc commented on index.html:
```diff
@@ -1401,9 +1401,47 @@ <h2>Core Vocabulary Definitions</h2>
-     used throughout a TD document is defined within an object as a member with 
+     used throughout a TD document for human-readable strings outside of a <a href="#multilanguage">MultiLanguage</a> object
+     is also defined within an object as a member with 
      a key <code>@language</code> whose value MUST be a string representing a [[!BCP47]] 
```
Should be "the name `@language` whose ... language tag."

mkovatsc replied:
Talking about "member" would make it a serialization assertion.

mmccool replied:
I'll see what I can do to clean this up.

mmccool replied:
Done.

mkovatsc commented on index.html:
```diff
@@ -1401,9 +1401,47 @@ <h2>Core Vocabulary Definitions</h2>
      a key <code>@language</code> whose value MUST be a string representing a [[!BCP47]] 
      language code.</span>
+     <span class="rfc2119-assertion" id="td-context-default-language-last">
+     The object used to define the default language MUST be the last object in
```
MUST be the last _element_ in

mkovatsc replied:
However, this appears to be a serialization issue to me...

mmccool replied:
See also the next item.  I can remove this if it would cause problems with TD generators based on a JSON-LD library.

mmccool replied:
Removed.

mkovatsc commented on index.html:
```diff
@@ -1401,9 +1401,47 @@ <h2>Core Vocabulary Definitions</h2>
+     </span>
+     <span class="rfc2119-assertion" id="td-context-default-language-separate">
+     The object used to define the default language MUST NOT be combined with any other
+     objects defining context extensions (additional vocabulary terms).
```
This is an issue when a JSON-LD library is used to serialize a TD. I would rather relax this. (Also applies to the position in the array...)

mmccool replied:
Hmmm.... good point, I guess.   And to the other point.   Well, I thought I would throw this out there, I'm not married to it.

mkovatsc replied:
Lucky you :D
Good remove it.

mmccool replied:
Removed.   I also removed the assertion that was an aim to avoid duplicates.   However, this is a serialization issue and should be handled by JSON-LD.

mkovatsc commented on index.html:
```diff
@@ -1401,9 +1401,47 @@ <h2>Core Vocabulary Definitions</h2>
+     <li><span class="rfc2119-assertion" id="td-context-default-language-direction-default">
+     If no default language is given, the base text direction MUST be assumed to be LTR (left-to-right).
+     </span>  This implies that if the language used in human readable text uses a script
+     that is written RTL (right-to-left), the default language needs to be explicitly specified.</li>
```
the default language, _potentially including a script subtag_, needs to be ...

mmccool replied:
OK.

mmccool replied:
Phrase added.

mkovatsc commented on index.html:
```diff
@@ -1401,9 +1401,47 @@ <h2>Core Vocabulary Definitions</h2>
+     that is written RTL (right-to-left), the default language needs to be explicitly specified.</li>
+     <li><span class="rfc2119-assertion" id="td-context-default-language-direction-independence">
+     Each human-readable string MUST be processed independently, starting from the base
+     direction defined for the TD.</span></li>
```
Where does this come from? Does it mean the direction has to be computed every time, although the default language cannot have changed?

mmccool replied:
No, it means you can't carry forward changes in direction from one string to another, or take the order that strings in the document appear as part of the processing.   Each string is processed independently, starting from the same base direction.  Which is what I said.

mmccool replied:
This actually relates to some points in the internationalization questionnaire.   Anyway, it seems obvious, but I added it to make SURE that no one does this.   You could imagine someone doing something silly like processing the entire JSON document as an string...

mkovatsc replied:
I easily understood the explanation with "can't carry forward changes in direction from one string to another", opposed to the assertion text... maybe we can improve accordingly

mkovatsc replied:
I easily understood the explanation with "can't carry forward changes in direction from one string to another", opposed to the assertion text... maybe we can improve accordingly.

This should go to Section 6.

mmccool replied:
I reworded the assertion a little but, but also added an "In other words..." statement after it to clarify.  Also moved to Section 6 since this is a serialization (or rather, a deserialization) concern.

mkovatsc commented on index.html:
```diff
@@ -1494,12 +1532,24 @@ <h2>Core Vocabulary Definitions</h2>
+<section><h3><code>MultiLanguage</code></h3><p>Container to provide human-readable text in different languages using language tags described in [[BCP47]]. See <a href="#titles-descriptions-serialization-json"></a> for example usages of this container in a Thing Description instance.</p>
 <p>
+<span class="rfc2119-assertion" id="td-context-ns-multilanguage-language-tag">
+Each language tag code used MUST be declared as vocabulary term member of this container 
```
Strike _member_ here

mmccool replied:
OK; consistency.

mkovatsc replied:
Victor updated the Model definitions. I think the terminology of this is now quite off. MultiLanguage is now a "Map", an "Object" with only "Terms", not "Vocabulary Terms" as names.

Each name of the <code>MultiLanguage</code> <a>Map</a> MUST be a language tag as defined in [[!BCP47]] (e.g., <code>en</code>, <code>de</code>, <code>ja</code>, <code>zh-Hans</code>, <code>zh-Hant</code>,
<code>az-Latn</code>, <code>az-Arab</code>).

mmccool replied:
Yeah, some of this text is not mine, but needs to be updated for consistency.  Done.  A few other related things also updated.

mkovatsc commented on index.html:
```diff
@@ -1494,12 +1532,24 @@ <h2>Core Vocabulary Definitions</h2>
+</span>
+<span class="rfc2119-assertion" id="td-context-ns-multilanguage-text-direction-script-subtag">
+If a language can be written in more than one script with different base directions 
+then a script subtag MUST be used so that the base direction can be inferred.
```
... MUST be _included_ so that ...
?

mmccool replied:
Sure, means basically the same thing, but I can make the change.

mmccool replied:
OK, I can see where "included" can make more sense since it is always part of the language tag.  Anyhow, done.

mkovatsc commented on index.html:
```diff
@@ -1494,12 +1532,24 @@ <h2>Core Vocabulary Definitions</h2>
+(e.g., <code>en</code>, <code>de</code>, <code>ja</code>, <code>zh-Hans</code>, <code>zh-Hant</code>,
+<code>az-Latn</code>, <code>az-Arab</code>). </span>
+<span class="rfc2119-assertion" id="td-context-ns-multilanguage-text-direction-infer">
+The base text direction MUST be inferred from the language tag code.
```
strike _code_

mkovatsc replied:
This is a good candidate for Section 6.

mmccool replied:
Code markup removed.   I also added Azeri as an example of where this is needed.  I am not sure this goes in Section 6, as it is not a serialization issue.   Text direction is how the text is displayed when it is rendered, i.e. how the data is to be interpreted.   Therefore you also need the text direction in the text's information model.

mmccool replied:
I left this in the MultiLanguage section, and as discussed below, removed a redundant assertion in Section 6.

mkovatsc commented on index.html:
```diff
@@ -1494,12 +1532,24 @@ <h2>Core Vocabulary Definitions</h2>
+In protocols such as HTTP that have an ability to negotiate content based on languages, 
+those terms (e.g. <code>description</code> and <code>title</code>) intended 
+for carrying texts in a default language SHOULD be used instead of <code>MultiLanguage</code> container
+when the texts are served as a result of such negotiation.
```
Only works for dynamically serialized TDs...

mmccool replied:
Actually, I don't think I changed this, it just got moved or the formatting got changed.   What you say is true, but I don't think it changes the assertion.

mkovatsc replied:
I think there should be a comment after the assertions that informs that this does not mean that Things should be able to dynamically generate their TD. This might be hard for constrained devices.

mmccool replied:
OK, I can add that comment.   SHOULD gives us some wiggle room.

mmccool replied:
I added the following sentence immediately after this assertion (it is not itself an assertion, just a note): "Note however that constrained devices may choose to not support dynamically-generated TDs and do not need to support this feature."

mmccool commented:
I agree there are a lot of issues that are serialization related.   On the other hand I think being specific about things like where the `@language` tag can appear in the `@context` will make things more consistent and easier to parse.  

I'll see if I can find a good place to move the serialization constraints that are more logical.   I can update the PR tomorrow... I have to anyway to resolve the conflicts now.

mkovatsc commented:
+1 to define the allowed location of `@language` and its meaning of default language in the model (Section 5).

The serialization-specific issues including the i18n aspect should go into Section 6. There I would rename the titles and descriptions section to "Human-readable Metadata Serialization".

mmccool commented:
I'm not convinced these text direction issues are serialization issues.  Text direction affects how the text is displayed, but does not affect how the bytes are ordered in storage or when communicated (which is what "serialization" means).   Given that, do you still feel I should move the points I left open above to Section 6?

mmccool commented:
I took a look at section 6 and noticed some new assertions had been added at some point that were redundant with the ones I had added under MultiLanguage, eg the use of language tags and inference of text direction.   However, I still assert that these are NOT serialization issues since the text direction does not affect how the data is stored, it affects how it is *interpreted*, so it's legitimately part of the information model.  However, redundancy is bad, so I removed the extra assertions (actually, they were tacked onto the end of another assertion and not given their own id, which was another problem...).   I did rename the section to "Human-Readable Metadata" and added a note about the fact the `@language` can be serialized in its own object or combined with others.

mmccool commented:
But HTTP language negotation is arguably part of serialization, so I did move that assertion to section 6.

mmccool commented:
Also moved assertion for "independent processing" to Section 6.

mmccool commented:
I think all the issues raised in the review have been addressed, and this is ready to merge.

mmccool commented:
I created an additional PR to add a definition for "supportLanguage":https://github.com/mmccool/wot-thing-description/pull/8.    This is a modification of this PR.

mkovatsc commented on index.html:
```diff
@@ -2968,26 +3000,47 @@ <h3>Data Schema Serialization</h3>
+  In protocols such as HTTP that have an ability to negotiate content based on languages,
+  terms (e.g. <code>description</code> and <code>title</code>) intended
+  for carrying texts in the default language SHOULD be used instead of <code>MultiLanguage</code> container
+  when the texts are served as a result of such negotiation.</span>
```
In protocols such as HTTP that have an ability to negotiate content based on languages,
TD documents SHOULD only include JSON members based on the default language
(e.g., <code>description</code> and <code>title</code>)
and no <code>MultiLanguage</code> objects,
while the <code>@language</code> member is omitted.

mmccool replied:
I am fine with adding an assertion to leave out the MultiLanguage objects.   I'm not convinced it is a good idea to leave out @language since if the TD gets forwarded somewhere it would be useful.  I think instead we should add an assertion that if it is included, it MUST be consistent with the negotiated default language.

mmccool replied:
I added the following two assertions; notice the first one mandates leaving OUT MultiLanguage objects which seems simpler.
* In contexts where the default language has been negotiated, TD documents SHOULD NOT include any MultiLanguage objects. 
* In contexts where the default language has been negotiated, a @language definition, if present, MUST be consistent with the negotiated default language.

mkovatsc replied:
Good

mkovatsc commented on index.html:
```diff
@@ -2968,26 +3000,47 @@ <h3>Data Schema Serialization</h3>
+  and do not need to support this feature.
+  </p>
+  <p>
+  As noted earlier, the base text direction used to display human-readable text is also inferred 
```
strike "also" before inferred

mmccool replied:
sure.

mkovatsc commented on index.html:
```diff
@@ -2968,26 +3000,47 @@ <h3>Data Schema Serialization</h3>
-
   <p>
   A TD snippet using <code>title</code> and <code>description</code> is 
   shown below. The default language is set to <code>"en"</code> through the
```
remove quotes around `en` in code tags.

mmccool replied:
Actually, elsewhere you had me remove the "code" markup around language tags, too.   We should be consistent.   It looks a little odd with nothing around it, and it is a string value.   I think the other places language tags are given as examples they should get the "code" treatment to be consistent.

mmccool replied:
OK, fixed... but I also went back and put the language tag examples in the "Multilanguage" section inside "code" markup to be consistent.

mkovatsc replied:
Sorry, I cannot remember saying that the code markup should be removed from language tag examples. As they are Terms, they should have code markup.

mkovatsc commented on index.html:
```diff
@@ -2968,26 +3000,47 @@ <h3>Data Schema Serialization</h3>
   A TD snippet using <code>title</code> and <code>description</code> is 
   shown below. The default language is set to <code>"en"</code> through the
-  <code>@language</code> member within the <code>@context</code>.
+  definition of the <code>@language</code> name within the <code>@context</code>.
```
<code>@language</code> JSON member within the <code>@context</code>.

mmccool replied:
fixed.

mkovatsc commented on index.html:
```diff
@@ -2968,26 +3000,47 @@ <h3>Data Schema Serialization</h3>
-  <code>@language</code> member within the <code>@context</code>.
+  definition of the <code>@language</code> name within the <code>@context</code>.
+  This definition can be serialized either in its own object or
+  combined with other definitions.
```
I would remove these two lines, as they only add confusion.

mmccool replied:
Eh, sure.

mkovatsc commented on index.html:
```diff
@@ -3018,17 +3071,20 @@ <h3><code>title(s)</code> and <code>description(s)</code></h3>
 
   <p>
   The <a>vocabulary terms</a> <code>titles</code> and <code>descriptions</code>
```
The <a>Vocabulary Terms</a> ...

mmccool replied:
Huh.  Not even sure that was my text, but I will fix it.

mmccool replied:
Upon searching for where you noted this issue I notice a bunch of other places where Vocabulary Terms was not upper-cased and/or anchored.   Since I assume this was supposed to be a global convention for special terminology I updated them all (at least the ones in index.template.html; there may be others pulled in as part of rendering).

mkovatsc commented on index.html:
```diff
@@ -3018,17 +3071,20 @@ <h3><code>title(s)</code> and <code>description(s)</code></h3>
   The <a>vocabulary terms</a> <code>titles</code> and <code>descriptions</code>
-  enable human-readable texts in multiple languages using language tags defined in [[!BCP47]].
+  enable human-readable texts in multiple languages using language tags defined in [[!BCP47]]
+  as described for the <code>MultiLanguage</code> Map.
```
... <a>Map</a>.

mmccool replied:
Not sure what you wanted here, however this was meant to be a reference back to the MultiLanguage section.   I updated it so it is an actual section reference; it now renders as the following:

The Vocabulary Terms titles and descriptions enable human-readable texts in multiple languages using language tags defined in [BCP47] as described in § 5.3.1.7 MultiLanguage.

mkovatsc commented on index.html:
```diff
@@ -3018,17 +3071,20 @@ <h3><code>title(s)</code> and <code>description(s)</code></h3>
   If a script subtag changes the text direction of the JSON string,
   it MUST be included in the name of the member.
-  </span>
+-->
```
Remove commented out lines (i.e., whole HTML comment)

mmccool replied:
Done.

mkovatsc commented:
@mmccool can you make sure that the default language also applies to the `name`? See comment in #630...

mmccool commented on index.html:
```diff
@@ -1400,9 +1400,34 @@ <h2>Core Vocabulary Definitions</h2>
-     used throughout a TD document is defined within an object as a member with 
-     a key <code>@language</code> whose value MUST be a string representing a [[!BCP47]] 
-     language code.</span>
+     used throughout a TD document for human-readable strings outside of a <a href="#multilanguage">MultiLanguage</a> object
```
So you are just saying expand "TD" to "Thing Description", and "object" to "instances", right?   I can do that.

mmccool replied:
Done: changed to
"used throughout a Thing Description for human-readable strings outside of MultiLanguage instances"

mkovatsc replied:
No, important is to drop "document", because "TD document" means serialized form.

mkovatsc replied:
Okay, is fixed correctly in the changes.

mmccool commented on index.html:
```diff
@@ -1400,9 +1400,34 @@ <h2>Core Vocabulary Definitions</h2>
+     include a script subtag so that an appropriate base direction can be inferred. 
+     </span></li>
+     <li><span class="rfc2119-assertion" id="td-context-default-language-direction-default">
+     If no default language is given, the base text direction MUST be assumed to be LTR (left-to-right).</span>
```
No problem there.  Done.

mmccool commented on index.html:
```diff
@@ -1400,9 +1400,34 @@ <h2>Core Vocabulary Definitions</h2>
+     is based on the following set of rules:
+     <ul>
+     <li><span class="rfc2119-assertion" id="td-context-default-language-direction-inference">
+     The base text direction MUST be inferred from the default language.
```
Well... this is for texts outside of MultiLanguage, as stated in the introductory paragraph.   However, it is true that this assertion is not standalone.   Also, if we add my supportLanguage extension, there will be another exception.   However, I can replace this with your sentence (and extend it again in my extended PR for supportLanguage...)

mmccool replied:
I changed this to
"Outside of  MultiLanguage instances the base text direction MUST be inferred from the default language tag."
and added the clarifying text after the list of rules:
"Rules inside MultiLanguage instances are similar but use the language tags given inside such instances instead of the default language tag."
Later, under MultiLanguage, another set of assertions is given.

mmccool commented on index.html:
```diff
@@ -1400,9 +1400,34 @@ <h2>Core Vocabulary Definitions</h2>
+     that is written RTL (right-to-left),
+     the default language, potentially including a script subtag, needs to be explicitly specified.</li>
+     </ul>
+     Consumers and producers of TDs should be aware of certain special cases that can arise
```
Ah, I see TD Processor is defined in Terminology to basically mean producer or consumer.  OK, done.

mmccool replied:
Also added an anchor to refer back to terminology.

mmccool commented on index.html:
```diff
@@ -1400,9 +1400,34 @@ <h2>Core Vocabulary Definitions</h2>
-     a key <code>@language</code> whose value MUST be a string representing a [[!BCP47]] 
-     language code.</span>
+     used throughout a TD document for human-readable strings outside of a <a href="#multilanguage">MultiLanguage</a> object
+     is also defined as a name 
```
Changed.

mmccool commented on index.html:
```diff
@@ -1493,13 +1518,20 @@ <h2>Core Vocabulary Definitions</h2>
 </section>
 <section><h3><code>VersionInfo</code></h3><p>Carries version information about the TD instance. If required, additional version information such as firmware and hardware version (term definitions outside of the TD namespace) can be extended here.</p><table class="def"><thead><tr><th><a>Vocabulary term</a></th><th>Description</th><th>Assignment</th><th>Type</th></tr></thead><tbody><tr class="rfc21 ...
-<section><h3><code>MultiLanguage</code></h3><p>Container to provide human-readable text in different languages using language tags described in [[BCP47]]. Each used language tags code MUST be decleared as vocabulary term member of this container (e.g., en, de, ja, zh-Hans, zh-Hant). See <a href="#titles-descriptions-serialization-json"></a> for example usages of this container in a Thing Descript ...
+<section><h3><code>MultiLanguage</code></h3><p>Container to provide human-readable text in different languages using language tags described in [[BCP47]]. See <a href="#titles-descriptions-serialization-json"></a> for example usages of this container in a Thing Description instance.</p>
```
Sure.   I also cleaned up the English in this sentence a little bit, specifically, it now says that this provides a *set* of texts.

mmccool commented on index.html:
```diff
@@ -1493,13 +1518,20 @@ <h2>Core Vocabulary Definitions</h2>
-when the texts are served as a result such negotiation.
-</span>
+<span class="rfc2119-assertion" id="td-context-ns-multilanguage-language-tag"> 
+Each name of the <code>MultiLanguage</code> Map MUST be a language tag, potentially including 
```
Done.   Actually I linked all instances of Map in this section.

mmccool commented on index.html:
```diff
@@ -2968,26 +3000,47 @@ <h3>Data Schema Serialization</h3>
   that default language is associated with the string values assigned to
   <code>title</code> and <code>description</code> within the TD
-  and to <code>name</code> in the Thing object.
+  and to <code>name</code> and <code>support</code> in the Thing object.
```
Yes, I think URLs just look at the order of bytes in memory, so the *rendering* direction does not matter.   So it's ok to say that the text direction of "support" is given by the default language even if the value is a URL.

mkovatsc replied:
We should make "support" of type anyURI and make it independent of the default language.

mmccool replied:
Well, I did suggest that as an alternative solution for "support".  Personally I think it is cleaner.  It does not resolve the default-language problem for "name", though.

mkovatsc replied:
Let's put "internationalization of support fields" under "next TD spec version".
Need to address "name" issue, see #630.

mmccool replied:
We can deal with this in a separate issue.

mmccool commented:
The intention is that the default language also applies to "name".   The text includes the following (in "Human-Readable Metadata"):
When a default language is specified by assigning a value to @language in the @context of a Thing object, that default language is associated with the string values assigned to title and description within the TD and to name and support in the Thing object.

mmccool commented:
Note I have not yet updated the supportLanguage sub-PR.   I will do that soon, I have to fix some conflicts now.

mkovatsc commented:
Arg, GitHub eats up the HTML tags I wrote in the comments.. thus confusion in places where I wanted you to add the markup such as anchoring to the definition...

mkovatsc commented:
We have a few comments on skipping on the supportLanguage feature.

The name must be defined to underlie the default language. The other way round, the name given must be in the default language (i.e., no English name when comments are in Arabic).

Support must be independent of default language (to become an anyURI).

mmccool commented:
+1 for making support an anyURI.   Probably we should create another issue for it, although I could certainly add that change to this PR.

The default-language-changing-upon-negotiation problem still exists for "name" but if I make a few small changes to my sub-PR (eg changing "supportLanguage" to "nameLanguage") it would be one possible solution.   Note in the issue I created I made a list of other possible solutions.   One simply one would be just to ignore the default language for "name" and use a strong-first rule to determine the text direction.   That is a little fragile (and is "not recommended" in the i18n guidelines) but would be simple.

My preference though would be to define "nameLanguage" and in (an update to...) my sub-PR.

mkovatsc commented:
See #630 comment. It does not make sense to introduce `names` and nameLanguage when we have `titles` already.

mkovatsc commented:
We should merge this soon, as it has lots over good and important changes!
Can you fix

> <code>title</code> and <code>description</code> within the TD
  and to <code>name</code> and <code>support</code> in the Thing object

to only `title` and `description`? Then we should be good unless there are other places that indicate `name`.

mkovatsc commented:
@mmccool Have you actively edited any of the ontology files or is it just render noise?

mmccool commented:
My edits were in index.template.html, validation/td-validation.ttl, and ontology/td.ttl.   Unfortunately there is some content hidden away in the ontology files that needed to be updated.  Everything else is render noise (specifically all of ontology/*.html and visualization/*.svg).

mmccool commented:
The last open issue in the reviews was about support.   I marked it as "resolved" but the true resolution will be to make it an anyURI.

I think this is ready to be merged now.

mmccool commented:
I will let you do the merge when you are ready.

mkovatsc commented:
Waiting for @sebastiankb as I am not officially an Editor of this draft...

mmccool commented:
Ah... well, I *am* an editor.  Let me go ahead and merge, then.
